### DISCUSSION TITLE
BLE broadcasting and observing

### DISCUSSION THREAD

**Discussion by hmillington** [pybricks/support#1297] [Q&A]
I'm about to experiment with this for GBC modules: e.g. have a module tell the one to the right of if that balls are coming and/or check that it's ready to receive them.

I note that the broadcast and observe channels are set when declaring the hub which causes me problems because I run the same code on every hub: the appropriate function for the module connected to it is determined by its 'signature' (which devices are connected to which ports) which of course is not known until after the hub has been declared.  Ideally the channels for the specific module would be set in that function.

So, is there, or will there be, a way to set them afterwards, something like 

hub.broadcast_channel = 0
hub,observe_channels=[1])

?

Or, is there a way to declare the hub, do some initialisation, then dispose of it before delclaring it again with different parameters?

I guess for simplicity each hub should listen on one channel and broadcast on another, but can they do both on one channel? And can multiple hubs broadcast on a single channel? I guess the answer to that is yes, but then a recipient would need to determine which one sent it somehow, presumably by setting something in the data, and there's potential for messages to get lost given that the observe function returns the last received data...

**Comment by BertLindeman** [comment]
Hi,

Not sure if this is intended to work, but ....

You could discriminate on the hubname.
This is a working example:

```python

from pybricks.hubs import ThisHub
from pybricks.tools import wait
from pybricks.parameters import Color

# Initialize the Inventor Hub
hub = ThisHub()
hub = ThisHub(observe_channels=[4])
hub = ThisHub(observe_channels=[6])
hub = ThisHub(observe_channels=[10])
if hub.system.name() == "hubB.3.3.0C1":
    hub = ThisHub(observe_channels=[1])

hub.light.on(Color.CYAN)
print("BlueTooth chip version", hub.ble.version(), "\n")


while True:
    # Receive broadcast from the other hub.
    data = hub.ble.observe(1)
    hub.light.on(Color.GREEN)
    if data is not None: 
        # hub.light.on(Color.BLUE)
        # print(" L" + str(len(data)) + str(data), end="")
        print("+", end="")

        start_volume = data[0]
        if start_volume == -1:
            hub.light.on(Color.RED)
            print("\n\tReceived stop command")
            raise SystemExit()
    else:
        hub.light.on(Color.ORANGE)
        print(".", end="")

    wait(100)

```

**Reply by hmillington** [reply]
That's very useful, thank you.

**Comment by dlech** [comment]
> but then a recipient would need to determine which one sent it somehow, presumably by setting something in the data

Right, this is exactly what the channel is intended to do so that you don't have to come up with your own solution.



> because I run the same code on every hub

Does it have to be this way? For example, you could put all of the common code in a `gbc_lib.py` file then make one script for each hub that specifies the channels specific to that hub.

```python
# gbc_hub_1.py

from gbc_lib import main

main(ch=1)
```

```python
# gbc_hub_2.py

from gbc_lib import main

main(ch=2)
```

```python
# gbc_lib.py

from pybricks.hubs import TechnicHub

def main(ch):
    hub = TechnicHub(broadcast_channel=ch, observe_channels=[1, 2, 3, 4, 5])
    ...
```

> it is determined by its 'signature' (which devices are connected to which ports) which of course is not known until after the hub has been declared

Why can't you determine the connected devices before declaring the hub? None of the sensor/motor classes require a hub object.

**Reply by hmillington** [reply]
Thank you both. 

I had assumed, incorectly it seems, that I needed to declare the hub before being able to tell what was connected to it. As that is not the case, then problem solved, thank you!

I run the same code on every hub for simplicity when setting up for a show: I don't need to label up hubs and remember which works which with module., I know a module will work with whichever one comes out the box next. In fact, I run the same code on Prime, Essential and Technic hubs to keep life even simpler so a module that needs just 2 ports can be run from any of them.

**Reply by laurensvalk** [reply]
Here is another variant for future possibilities:

You could have a very basic "menu" in your program to select a different "sub program". Even on the hubs with one button, you could toggle through a few colors to choose your program.

You could even save the selection in user storage so you'd have to do it only once.

**Comment by hmillington** [comment]
I have this working now, with three modules communcating their state, either ready, operating or discharging. 

Module #1 waits until #2 is ready to receive balls, which remains idle until some are forthcoming. Module three waits until #2 is discharing before springing into life. It does away with the need for light sensors on each module to detect the presence of balls, and also in some cases, gates that would otherwise be needed to hold balls back until the machine is ready.

I will make a video sometime...

Thanks everyone for your help.
